### PR TITLE
TTML: Fixed improper spacing between captions

### PR DESCRIPTION
**Pull request by discovery-avishekgulshan**
1. Fixed improper spacing between captions while converting TTML -> VTT
2. Converted TTML style attributes to pointers
3. Fixed Webvtt out not using parent style attributes in the "region" section

### CODE REVIEW

File: subtitles.go [hunk 1]
```diff
@@ -327,7 +339,9 @@ func (l Line) String() string {
 	for _, i := range l.Items {
 		texts = append(texts, i.Text)
 	}
-	return strings.Join(texts, " ")
+	t := strings.Join(texts, " ")
+	t = strings.ReplaceAll(t, "  ", " ")
```
Comment (asticode) [inline subtitles.go:343]: If you're trying to replace multi spaces into single spaces, I think you'd be better off using a regexp

File: ttml.go [hunk 2]
```diff
@@ -70,35 +70,48 @@ type TTMLInMetadata struct {
...
+}
+
+// StyleAttributes converts TTMLInStyleAttributes for region into a StyleAttributes
+func (s TTMLInStyleAttributes) regionStyleAttributes() (o *StyleAttributes) {
+	o = getStyleAttributes(s)
+	o.propagateTTMLRegionAttributes()
+	return
+}
+
+func getStyleAttributes(s TTMLInStyleAttributes) *StyleAttributes {
```
Comment (asticode) [inline ttml.go:113]: I'd rather rename this function to something more TTML-specific and more self-explanatory like `newStyleAttributesFromTTMLInStyleAttributes`

File: util/util.go [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package util
+
+func StringToPointer(str string) *string {
```
Comment (asticode) [inline util/util.go:3]: Can you remove this function and use [this one](https://github.com/asticode/go-astikit/blob/master/ptr.go#L41) instead (`astikit` is already used in the lib)

File: subtitles.go (hunk 1 above)
Comment (asticode) [inline subtitles.go:343]: Also, can you provide an example `ttml` that was having issue with that bit of code?

File: ttml.go [hunk 4]
```diff
@@ -394,7 +405,7 @@ func ReadFromTTML(i io.Reader) (o *Subtitles, err error) {
 			// New line decoded as a line break. This can happen if there's a "br" tag within the text since
 			// since the go xml unmarshaler will unmarshal a "br" tag as a line break if the field has the
 			// chardata xml tag.
-			for idx, li := range strings.Split(tt.Text, "\n") {
+			for idx, li := range strings.Split(tt.Text, "\\n") {
```
Comment (asticode) [inline ttml.go:408]: Can you provide an example `ttml` that was having issue with that bit of code?

File: ttml.go [hunk 5]
```diff
@@ -421,6 +432,15 @@ func ReadFromTTML(i io.Reader) (o *Subtitles, err error) {
 			}
 
 		}
+		//if multiple items in `l` are present without new line then adding space at end so it can become readable after concatenation
+		if len(l.Items) > 0 {
```
Comment (asticode) [inline ttml.go:436]: Can you provide an example `ttml` that was having issue with that bit of code?

**Review by asticode — COMMENTED** [review]
I know this is only a draft, but I thought it could still be productive to share my early comments

File: ttml.go (hunk 2 above)
Comment (discovery-avishekgulshan) [reply ttml.go:113]: done

File: util/util.go (hunk 3 above)
Comment (discovery-avishekgulshan) [reply util/util.go:3]: done

File: subtitles.go (hunk 1 above)
Comment (discovery-avishekgulshan) [reply subtitles.go:343]: I used the same ttml in testdata i.e. example-in.ttml

While converting from TTML -> VTT 
TTML: 
` <p xml:id="sub_2" begin="00:02:04.080" end="00:02:07.120" region="region_2">
                MAN:
                <tt:br/>
                How did we <span style="style_1" color="green">end up</span> here?
            </p>`
 VTT:
Expected: How did we end up here?
Actual: How did weend uphere?

File: ttml.go (hunk 4 above)
Comment (discovery-avishekgulshan) [reply ttml.go:408]: I used testdata>example-in.ttml, added '\n' n tested, it was not working.

File: ttml.go (hunk 5 above)
Comment (discovery-avishekgulshan) [reply ttml.go:436]: I used testdata/example-in.ttml

After adding this fix, SRT, and STL output added this additional space while joining the texts, so I added code to replace extra space while joining the Text for SRT and STL.

File: subtitles.go (hunk 1 above)
Comment (discovery-avishekgulshan) [reply subtitles.go]: done

**Comment by asticode** [comment]
I'm OK with `Converted TTML style attributes to pointers` and `Fixed Webvtt out not using parent style attributes in the "region" section` but I'm still wondering what was wrong with `Fixed improper spacing between captions while converting TTML -> VTT`.

Can you share one or more `ttml` that were causing issue before your changes?

**Comment by asticode** [comment]
Never mind, I got it

File: subtitles.go [hunk 6]
```diff
@@ -327,7 +340,9 @@ func (l Line) String() string {
 	for _, i := range l.Items {
 		texts = append(texts, i.Text)
 	}
-	return strings.Join(texts, " ")
+	t := strings.Join(texts, " ")
```
Comment (asticode) [inline subtitles.go:343]: This feels hacky, I think if we implement the rule `trim spaces on input, add in-between spaces on output`, we don't need this logic anymore

File: ttml.go (hunk 5 above)
Comment (asticode) [inline ttml.go:436]: I think if we implement the rule `trim spaces on input, add in-between spaces on output`, we don't need this logic anymore.

However you can add this logic on output [here](https://github.com/asticode/go-astisub/blob/master/ttml.go#L643) to make sure we add `lineItem.Text + ""` when `idx < len(item.Lines) -1`

File: ttml.go (hunk 4 above)
Comment (asticode) [reply ttml.go:408]: What do you mean `it was not working`? What did you expect to see?

**Review by asticode — CHANGES_REQUESTED** [review]
I think implementing the rule `trim spaces on input, add in-between spaces on output` is a better idea.

It would mean making sure to:

- trim space [here](https://github.com/asticode/go-astisub/blob/master/srt.go#L107) and [here](https://github.com/asticode/go-astisub/blob/master/ssa.go#L1017) (I think spaces are already trimmed on input for all other formats)
- add in-between spaces [here](https://github.com/asticode/go-astisub/blob/master/ssa.go#L930) (this is a bug here, you need to replace `""` with `" "`), [here](https://github.com/asticode/go-astisub/blob/master/ttml.go#L643) and [here](https://github.com/asticode/go-astisub/blob/master/webvtt.go#L400) (I think in-between spaces are already added for all other formats)

File: ttml.go (hunk 4 above)
Comment (discovery-avishekgulshan) [reply ttml.go:408]: The code is supposed to spilt the string on '\n' but it was not splitting so I added this extra '\' and now it's working.

File: stl.go [hunk 7]
```diff
@@ -673,7 +673,7 @@ func stlVerticalPositionFromStyle(sa *StyleAttributes) int {
 }
 
 func (li LineItem) STLString() string {
-	rs := li.Text
+	rs := strings.TrimSpace(li.Text)
```
[the remark this answers is not in the document]
Comment (discovery-avishekgulshan) [reply stl.go]: Thanks for your help, did all the changes

File: subtitles.go (hunk 6 above)
Comment (discovery-avishekgulshan) [reply subtitles.go]: yep, did felt hacky. Removed it

File: ttml.go (hunk 4 above)
Comment (asticode) [reply ttml.go:408]: Mmmm I'm really not sure `.ttml` text should be split on string `\n` 🤔 I think it should only be split on line returns.

Can you share some official docs or RFC that provide the same information as you?

**Review by asticode — CHANGES_REQUESTED** [review]
You forgot to add in-between space [here](https://github.com/asticode/go-astisub/blob/master/ttml.go#L643)

File: ttml.go (hunk 4 above)
Comment (discovery-avishekgulshan) [reply ttml.go:408]: I also think the same, but the code was written to split on new line characters, and during my testing, this code to split was not working, so I added this extra slash.
Anyway, removing it.

**Comment by asticode** [comment]
Can you juste remove changes to `go.sum` from this PR?

Then I'll merge the PR.

**Comment by discovery-avishekgulshan** [comment]
> Can you juste remove changes to `go.sum` from this PR?
> 
> Then I'll merge the PR.

go.sum automatically got updated when I ran go mod download. Trying to replace with the original file.

**Comment by asticode** [comment]
Thanks for sticking with this PR.

FYI I've created a `v0.13.0` tag.

Cheers

**Pull request merged**